Lift trivial Ifs

ThrudPrimrose opened the pull request:
Dead control flow elimination does not simplify by trying to evaluate trivial if conditions. I wrote a pass that does that:

It tries to eval if branches that are trivially true and then lifts the inner CFG upwards:
<img width="442" height="699" alt="image" src="https://github.com/user-attachments/assets/913a70e2-bee1-42e3-afbf-abd8be385715" />
Becomes:
<img width="455" height="699" alt="image" src="https://github.com/user-attachments/assets/5ad125e3-b4cf-4cf1-b040-d43c4a6bda6c" />


It supports branches of type:
```
if (cond)
```

and (empty else):
```
if(cond) 
...
else:
....
```

Example:
<img width="738" height="635" alt="image" src="https://github.com/user-attachments/assets/43c9bce7-41e7-4894-a4d1-876170400ab1" />
Becomes:
<img width="699" height="680" alt="image" src="https://github.com/user-attachments/assets/d08da18c-0825-4afc-8956-cad007b3e6d9" />

ThrudPrimrose committed:
Lift trivial fi

ThrudPrimrose committed:
Run precommit hook on code

ThrudPrimrose committed:
Fix union issue

ThrudPrimrose committed:
Rm rogue test

acalotoiu reviewed (approved):
LGTM!

tbennun reviewed (changes requested):
minor change

ThrudPrimrose committed:
Use symbolic eval instead ast node eval

ThrudPrimrose committed:
Fixes

ThrudPrimrose committed:
Merge branch 'main' into lift_trivial_if

ThrudPrimrose committed:
Fix import

ThrudPrimrose committed:
Precommit

tbennun commented on dace/transformation/passes/lift_trivial_if.py:
```diff
@@ -26,14 +27,56 @@ def should_reapply(self, modified: ppl.Modifies) -> bool:
+        }
+        all_labels = set()
+
+        def _find_new_name(cfg: ControlFlowRegion) -> str:
```
Why not use data.find_new_name(cfg_label, all_labels)?

ThrudPrimrose replied:
Now, having a pass to replace those integer comparisons with proper booleans seem good to me, still I think I should have some fallback code just in case.

tbennun reviewed (changes requested):
More comments, mostly about reusing existing infrastructure

ThrudPrimrose commented on dace/transformation/passes/lift_trivial_if.py:
```diff
@@ -26,14 +27,56 @@ def should_reapply(self, modified: ppl.Modifies) -> bool:
+            result = symbolic.evaluate(expr=dace.symbolic.SymExpr(pystring), symbols=dict())
             return bool(result) is val
         except Exception as e:
+            print(e)
```
I guess no

ThrudPrimrose commented on dace/transformation/passes/lift_trivial_if.py:
```diff
@@ -78,16 +126,32 @@ def _detect_trivial_ifs_and_rm_cfg(self, graph: Union[ControlFlowRegion, SDFG],
         # We might now have trivial control flow blocks at top level, apply in fixpoint
+        rmed_count = self._detect_and_remove_top_level_trivial_ifs(graph)
         local_rmed_count = rmed_count
+        graph.sdfg.validate()
```
I didn't know it was default. I usually sprinkle validate calls to confirm correctness during the development and then I just don't remove them again as they wouldn't hurt. Performance wise they are completely unnecessary So, I will remove them

phschaad committed:
Merge branch 'main' into lift_trivial_if

ThrudPrimrose committed:
Address review comments + add LiftTrivialIf to SimplifyPass

ThrudPrimrose committed:
Refactor components to utility class

ThrudPrimrose committed:
Merge branch 'main' into lift_trivial_if_review

ThrudPrimrose committed:
Merge branch 'main' into lift_trivial_if

ThrudPrimrose committed:
Clean-up PR

tbennun commented on dace/sdfg/construction_utils.py:
```diff
@@ -0,0 +1,45 @@
+    Moves a branch of a conditional block up in the control flow graph (if_block) (CFG),
+    replacing the conditional with the selected branch (body_to_take), discarding
+    the conditional check and other branches. Connects the body to previous in and out neighbors of the CFG.
+    """
```
This is not formatted in the Sphinx documentation format

tbennun commented on dace/sdfg/construction_utils.py:
```diff
@@ -0,0 +1,45 @@
+# Copyright 2019-2026 ETH Zurich and the DaCe authors. All rights reserved.
```
Needs a file docstring stating what is the purpose of this file and why this cannot just be in transformation helpers.

Honestly I would move this function there.

ThrudPrimrose replied:
I moved the function to the helpers file.

tbennun commented on dace/transformation/passes/lift_trivial_if.py:
```diff
@@ -0,0 +1,152 @@
+
+
+@transformation.explicit_cf_compatible
+class LiftTrivialIf(ppl.Pass):
```
needs docstring (also to appear properly in vscode!)

ThrudPrimrose replied:
I added the docstring.

tbennun commented on tests/passes/lift_trivial_if_test.py:
```diff
@@ -0,0 +1,253 @@
+# expression (Function/Indexed/Symbol). bool() of such an expression is
+# truthy, which would mis-classify dynamic dataflow conditions like
+# `A[0]` as trivially true.
+dynamic_runtime_cond = ["A[0]", "tmp_r[0]", "x", "x[0] + 1", "A[i, j]"]
```
nit: I'd use all-caps variable names such as `ALWAYS_TRUE` to ensure this is a test constant. Maybe even `_ALWAYS_TRUE` so it is not visible by default.

ThrudPrimrose replied:
Makes sense, I have updated the variable names.

tbennun reviewed (changes requested):
Looks great, minor comments only

ThrudPrimrose committed:
lift_trivial_if: style sweep + move helper into transformation.helpers

Module docstrings on the pass file. Class docstring on LiftTrivialIf. Move
move_branch_cfg_up_discard_conditions into dace.transformation.helpers and
drop the dace.sdfg.construction_utils file (it held only that one helper).
Sphinx docstring on the moved function.

Tests: rename module-level fixture lists to ALL_CAPS with _ prefix
(_ALWAYS_TRUE, _ALWAYS_FALSE, _CANT_EVAL, _DYNAMIC_RUNTIME_COND). Module
docstring on the test file. Sphinx docstrings on every _get_*_sdfg helper
and every test_* function.

ThrudPrimrose committed:
Bump up copyright, check comments

ThrudPrimrose committed:
Fix Sphinx inline-literal backticks in docstrings

ThrudPrimrose committed:
Drop single-backtick roles; use double-backtick literals

ThrudPrimrose committed:
Flank double-backtick literals with spaces; pure-ASCII docstrings

ThrudPrimrose committed:
Reconcile backticks to repo/Sphinx convention (no forced pre-punctuation space)

tbennun commented:
cscs-ci run
